doc/crypto/overview/conventions.rst

```diff
@@ -370,8 +370,9 @@ In some environments, an application can make calls to the |API| in
 separate threads. In such an environment, *concurrent calls* are two or more
 calls to the API whose execution can overlap in time.
 
-Concurrent calls are performed correctly, as if the calls were executed in
-sequence, provided that they obey the following constraints:
+The result of two or more concurrent calls must be consistent with the same set
+of calls being executed sequentially in some order, provided that they obey the
```

Comment:
This new statement is definitely an improvement. However, we want to specifically allow situations in which a call fails with a resource-exhaustion error where it is impossible for such a failure to occur in sequential execution. There is an example of this scenario here https://github.com/Mbed-TLS/mbedtls/pull/8618#issuecomment-1858264256. My interpretation of the suggested statement (there exists some valid ordering of the calls where for all `k`: [call `k` succeeds in the sequential execution] iff [call `k` succeeds in the concurrent execution]) doesn't permit this behaviour.

It is difficult to word the desired behaviour concisely. One suggestion is to modify the statement to say "The result (excluding resource exhaustion errors) of two or more..." and then add some text after the constraint list saying something like:
```
It is always permissible for a concurrent call to fail with a
resource exhaustion error (PSA_ERROR_INSUFFICIENT_MEMORY,
PSA_ERROR_INSUFFICIENT_STORAGE, PSA_ERROR_COMMUNICATION_FAILURE).
```

Reply:
I see your point. In some ways I was trying to avoid explicitly complicating the rule by discussing resource exhaustion, in the use of 'consistent with' - which is intentionally not 'the same result as if'. However, that probably depends on a nuanced reading of 'consistent with', which is not ideal for a spec that is not in a reader's primary language.

I would also prefer to avoid getting mired in description of side-effects on state that is explicitly part of the API programming model. (For example, keys in the key store and operation objects). Although that could clarify the intent of the concurrency rule - that failures due to interactions with non-API state (such as memory, storage etc) are not subject to the linearizability requirement - this gets very technical.

In practice - given the existing rules about undefined behavior if a key is destroyed concurrently with a call that uses it, and concurrent use of an operation object - the only API-state changing calls that are in scope are ones that create a key, and their interaction with other key creation or key usage calls on the same key identifier.

So we could either try and improve the clarification of the general rule. Or instead provide explicit guidance for the interaction of key creation and key usage APIs when called concurrently?

Reply:
However, I quite like the idea of trying to note the general caveat that errors resulting from resource constraints can arise in concurrent execution that do not arise in sequential execution.

Reply:
I think the mental 'trick' to understand what matters in the concurrency requirement is that the rule only cares about the behavior of the calls in relation to the specification-defined programming model. Therefore 'consistent with' does not mean 'the same as a sequential execution in the same implementation', but 'consistent with a valid sequential execution of the abstract programmer's model' - the abstract model does not provide or require guarantees about when resources might or might not be exhausted, whereas an implementation can provide resource-exhaustion-free guarantees for some calls.

However, that trick is not obvious to an implementer - they are just thinking about their own implementation.

Reply:
> I see your point. In some ways I was trying to avoid explicitly complicating the rule by discussing resource exhaustion, in the use of 'consistent with' - which is intentionally not 'the same result as if'. However, that probably depends on a nuanced reading of 'consistent with', which is not ideal for a spec that is not in a reader's primary language.
> 
> I would also prefer to avoid getting mired in description of side-effects on state that is explicitly part of the API programming model. (For example, keys in the key store and operation objects). Although that could clarify the intent of the concurrency rule - that failures due to interactions with non-API state (such as memory, storage etc) are not subject to the linearizability requirement - this gets very technical.
> 
> In practice - given the existing rules about undefined behavior if a key is destroyed concurrently with a call that uses it, and concurrent use of an operation object - the only API-state changing calls that are in scope are ones that create a key, and their interaction with other key creation or key usage calls on the same key identifier.
> 
> So we could either try and improve the clarification of the general rule. Or instead provide explicit guidance for the interaction of key creation and key usage APIs when called concurrently?



I agree that we should avoid discussing key stores and the like.

If we only provide guidance for interactions involving key creation then it is unclear whether we allow non-linearizable resource exhaustion errors in other cases where it is desirable. For example, the usage of persistent keys that are valid but not currently in memory.

In my opinion it is better to be general, while still making clear that non-linearizable resource exhaustion errors are permitted where calls would otherwise succeed.

Reply:
> I think the mental 'trick' to understand what matters in the concurrency requirement is that the rule only cares about the behavior of the calls in relation to the specification-defined programming model. Therefore 'consistent with' does not mean 'the same as a sequential execution in the same implementation', but 'consistent with a valid sequential execution of the abstract programmer's model' - the abstract model does not provide or require guarantees about when resources might or might not be exhausted, whereas an implementation can provide resource-exhaustion-free guarantees for some calls.
> 
> However, that trick is not obvious to an implementer - they are just thinking about their own implementation.

By looking at the proposed constraints with this perspective I now see how the statement works. Your last sentence is true: as an implementer I first read the constraints from the implementation-fixed perspective you describe.

Reply:
I've update the PR with some organising structure (so it is clear which text relates to the 'consistency' rule), and added the resource-related error caveat/relaxation.

Reply:
Just thought: I should probably remove the 'Note that...' from the caveat, as it makes it appear informative, rather that normative and providing clear permission to an implementer?

Reply:
I am now happy to approve these changes, I think it now reads clearly and correctly - I agree that removing "Note that" is a good idea.